Advanced customization guide (#5066)
adobe/react-spectrum: packages/react-aria-components/docs/advanced.mdx

This example shows a `FieldGroup` component that renders a group of text fields. The entire group can be marked as disabled via the `isDisabled` prop, which is passed to all child text fields via the `TextFieldContext` provider.

```tsx
import {TextFieldContext} from 'react-aria-components';

interface FieldGroupProps {
  children?: React.ReactNode,
  isDisabled?: boolean
}

function FieldGroup({children, isDisabled}: FieldGroupProps) {
  return (
    /*- begin highlight -*/
    <TextFieldContext.Provider value={{isDisabled}}>
    {/*- end highlight -*/}
      {children}
    </TextFieldContext.Provider>
  );
}
```

Any `TextField` component you place inside a `FieldGroup` will automatically receive the `isDisabled` prop from the group, including those that are deeply nested inside other components.

```tsx
<FieldGroup isDisabled={isSubmitting}>
  <MyTextField label="Name" />
  <MyTextField label="Email" />
  <CreditCardFields />
</FieldGroup>
```

The contexts consumed by each component are listed in the Advanced Customization section of their documentation.

### Slots

> Some patterns include multiple instances of the same component. These use the `slot` prop to distinguish each instance. Separate props can be sent to each slot by providing an object with keys for each slot name.

Reviewer: Do we need to explain this concept? I know we have a page already https://react-spectrum.adobe.com/react-spectrum/layout.html#slots but its in RSP, is that enough?

This example shows a `Stepper` component with slots for its increment and decrement buttons.

```tsx
function Stepper({children}) {
  let [value, setValue] = React.useState(0);

  return (
    <ButtonContext.Provider
      value={{
        slots: {
          increment: {
            onPress: () => setValue(value + 1)
          },
          decrement: {
            onPress: () => setValue(value - 1)
          }
        }
      }}>
      {children}
    </ButtonContext.Provider>
  );
}

<Stepper>
  <Button slot="increment">⬆</Button>
  <Button slot="decrement">⬇</Button>
</Stepper>
```

The slots supported by each component are shown in the Anatomy section of their documentation.

### Provider

> The `Provider` component is a utility that makes it easier to provide multiple React contexts without manually nesting them. Pass pairs of contexts and values as an array to the `values` prop.

Reviewer: ```suggestion
The `Provider` component is a utility that makes it easier to provide multiple React contexts without manually nesting them. This can be achieved by passing pairs of contexts and values as an array to the `values` prop.
```

#### useContextProps

The <TypeLink links={docs.links} type={docs.exports.useContextProps} /> hook merges the local props and ref with the ones provided via context by a parent component. The local props always take precedence over the context values (following the rules documented in [mergeProps](mergeProps.html)).

```tsx
import type {LabelProps} from 'react-aria-components';
import {LabelContext, useContextProps} from 'react-aria-components';

const MyCustomLabel = React.forwardRef(
  (props: LabelProps, ref: React.ForwardedRef<HTMLLabelElement>) => {
    // Merge the local props and ref with the ones provided via context.
    /*- begin highlight -*/
    [props, ref] = useContextProps(props, ref, LabelContext);
    /*- end highlight -*/

    // ... your existing Label component
    return <label {...props} ref={ref} />;
  }
);
```

Since it consumes from `LabelContext`, `MyCustomLabel` can be used within any component, instead of the builtin React Aria Components `Label`.

```tsx
<TextField>
  {/*- begin highlight -*/}
  <MyCustomLabel>Name</MyCustomLabel>
  {/*- end highlight -*/}
  <Input />
</TextField>
```

#### useSlottedContext

> To consume a context without merging with existing props, use the <TypeLink links={docs.links} type={docs.exports.useSlottedContext} /> hook. This works just like React's `useContext` but also accepts an optional slot argument to identify which slot name to consume.

Reviewer: ```suggestion
To consume a context without merging with existing props, use the <TypeLink links={docs.links} type={docs.exports.useSlottedContext} /> hook. This works like React's `useContext` and also accepts an optional slot argument to identify which slot name to consume.
```

### Setup

As described [above](#contexts), each React Aria component exports a corresponding context. You can build a custom implementation of a component using Hooks by consuming from the relevant context with <TypeLink links={docs.links} type={docs.exports.useContextProps} />.

This example shows how a custom checkbox could be set up using `CheckboxContext` from `react-aria-components` and the [useCheckbox](useCheckbox.html) hook from `react-aria`.

```tsx
import type {CheckboxProps} from 'react-aria-components';
import {CheckboxContext, useContextProps} from 'react-aria-components';
import {useToggleState} from 'react-stately';
import {useCheckbox} from 'react-aria';

const MyCheckbox = React.forwardRef((props: CheckboxProps, ref: React.ForwardedRef<HTMLInputElement>) => {
  // Merge the local props and ref with the ones provided via context.
  ///- begin highlight -///
  [props, ref] = useContextProps(props, ref, CheckboxContext);
  ///- end highlight -///

  // Follow the hook docs...
  let state = useToggleState(props);
  let {inputProps} = useCheckbox(props, state, ref);
  return <input {...inputProps} ref={ref} />;
});
```

> Since `MyCheckbox` consumes from `CheckboxContext`, it can be used within other React Aria Components in place of the default `Checkbox` component, such as within a [Table](Table.html) or [GridList](GridList.html). This lets you replace only the checkbox with a custom implementation, without rewriting all other React Aria Components you might use it in.

Reviewer: ```suggestion
Since `MyCheckbox` consumes from `CheckboxContext`, it can be used within other React Aria Components in place of the default `Checkbox` component, such as the ones found within a [Table](Table.html) or [GridList](GridList.html). This lets you specifically customize the checkbox without rewriting all other React Aria Components you might use.
```

```tsx
import type {CheckboxProps} from 'react-aria-components';
import {CheckboxContext, useContextProps} from 'react-aria-components';
import {useToggleState} from 'react-stately';
import {useCheckbox} from 'react-aria';

const MyCheckbox = React.forwardRef((props: CheckboxProps, ref: React.ForwardedRef<HTMLInputElement>) => {
  // Merge the local props and ref with the ones provided via context.
  ///- begin highlight -///
  [props, ref] = useContextProps(props, ref, CheckboxContext);
  ///- end highlight -///

  // Follow the hook docs and implement your customizations...
  let state = useToggleState(props);
  let {inputProps} = useCheckbox(props, state, ref);
  return <input {...inputProps} ref={ref} />;
});
```

Since `MyCheckbox` consumes from `CheckboxContext` it can be used within other React Aria Components in place of the built-in `Checkbox`, such as within a [Table](Table.html) or [GridList](GridList.html). This lets you provide a custom checkbox implementation without rewriting all other React Aria Components you might use it in.

```tsx
<GridList>
  <Item>
    {/*- begin highlight -*/}

>     <MyCustomCheckbox slot="selection" />

Reviewer 2: ```suggestion
    <MyCheckbox slot="selection" />
```
